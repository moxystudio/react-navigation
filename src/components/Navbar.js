import React, { useState, useMemo } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import { useDocumentScrollThrottled, useNavigation } from '../hooks';

const MINIMUM_SCROLL = 120;
const TIMEOUT_DELAY = 400;

const Navbar = ({ className, LinkComponent, navigationItems, placement }) => {
    const [shouldHideHeader, setShouldHideHeader] = useState(false);
    const [shouldShrinkHeader, setShouldShrinkHeader] = useState(false);

    const { toggleDrawer } = useNavigation();

    useDocumentScrollThrottled(({ previousScrollTop, currentScrollTop }) => {
        const isScrolledDown = previousScrollTop < currentScrollTop;
        const isMinimumScrolled = currentScrollTop > MINIMUM_SCROLL;

        setShouldShrinkHeader(isScrolledDown);

        setTimeout(() => {
            setShouldHideHeader(isScrolledDown && isMinimumScrolled);
        }, TIMEOUT_DELAY);
    });

    const containerClassName = useMemo(
        () =>
            classNames(
                'react-navigation__container',
                placement,
                {
                    hide: shouldHideHeader,
                    shrink: shouldShrinkHeader,
                },
                className,
            ),
        [placement, shouldHideHeader, shouldShrinkHeader, className],
    );

    const navbarClassName = useMemo(() => classNames('react-navigation__navbar', [placement]), [placement]);

    return (
        <div className={ containerClassName }>
            <nav className={ navbarClassName }>
                <div>
                    <LinkComponent href="/">
                        <a>next-with-moxy</a>
                    </LinkComponent>
                </div>
                <ul>
                    {navigationItems.map(({ href, text }, idx) => (
                        <li key={ idx }>
                            <LinkComponent href={ href }>
                                <a>{text}</a>
                            </LinkComponent>
                        </li>
                    ))}
                </ul>
            </nav>
            <button onClick={ toggleDrawer }>boom</button>
        </div>
    );
};

Navbar.propTypes = {
    className: PropTypes.string,
    LinkComponent: PropTypes.func.isRequired,
    navigationItems: PropTypes.arrayOf(
        PropTypes.exact({
            href: PropTypes.string,
            text: PropTypes.string,
        }),
    ).isRequired,
    placement: PropTypes.oneOf(['top', 'right', 'bottom', 'left']),
};

Navbar.defaultProps = {
    placement: 'top',
};

export default Navbar;
