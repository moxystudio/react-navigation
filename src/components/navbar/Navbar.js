import React, { useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import useBehaviour, { behaviours } from './utils/use-behaviour';

const navbarClassName = 'react-navigation_navbar';

const Navbar = ({ placement, behaviour, children, className }) => {
    const navbarRef = useRef(null);
    const [behaviourData, setBehaviourProps] = useBehaviour({ navbarRef, behaviour, placement });

    useEffect(() => {
        setBehaviourProps({ navbarRef, behaviour, placement });
    }, [setBehaviourProps, navbarRef, behaviour, placement]);

    const navbarClassNames = classNames(
        navbarClassName,
        `${navbarClassName}-${placement}`,
        behaviourData.className,
        className,
    );

    return (
        <nav
            { ...behaviourData.attributes }
            ref={ navbarRef }
            className={ navbarClassNames }>
            { children }
        </nav>
    );
};

Navbar.propTypes = {
    placement: PropTypes.oneOf(['top', 'right', 'bottom', 'left']),
    behaviour: PropTypes.oneOf(Object.keys(behaviours)),
    children: PropTypes.any,
    className: PropTypes.string,
};

Navbar.defaultProps = {
    placement: 'top',
};

export default Navbar;
