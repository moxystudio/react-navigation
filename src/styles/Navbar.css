@import "./variables.css";
@import "./custom-medias.css";

.react-navigation__container {
    position: fixed;
    top: 0;
    left: 0;
    width: 100vw;
    height: 100px;
    z-index: calc(var(--z-index-base) + 1);
    transform: translateY(0);
    display: flex;
    align-items: center;
    background-color: var(--color-rose-gold);
    color: var(--color-black);
    transition:
        height var(--duration-standard),
        transform var(--duration-standard);

    &.hide {
        transform: translateY(-110%);
    }

    &.shrink {
        height: 4rem;
    }

    & button {
        display: none;
        margin-right: 25px;
    }

    & a {
        color: inherit;
    }

    & > .react-navigation__navbar {
        width: 100%;
        margin: 0 25px;

        display: flex;
        justify-content: space-between;
        align-items: center;

        & ul {
            padding: 0;
            display: flex;
            list-style-type: none;

            & li:not(:last-child) {
                padding-right: 1.5rem;
            }
        }
    }
}

@media (--layout-lte-small) {
    .react-navigation__container {
        & button {
            display: inline-block;
        }

        & > .react-navigation__navbar {
            & ul {
                display: none;
            }
        }
    }
}
