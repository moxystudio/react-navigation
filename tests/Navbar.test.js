import React from 'react';
import { render, screen } from '@testing-library/react';

import Navbar from '../src/components/Navbar';
import AppTree from './AppTree';

const defaultProps = {
    placement: 'top',
    navigationItems: [
        { href: '/lorem', text: 'Lorem' },
        { href: '/ipsum', text: 'Ipsum' },
    ],
    LinkComponent: jest.fn(({ children, href }) => (
        <a href={ href }>
            {children.props.children}
        </a>
    )),
};

const renderWithProps = (props = {}, Component = Navbar) => render(
    <AppTree>
        <Component { ...defaultProps } { ...props } />
    </AppTree>,
);

describe('Navbar', () => {
    it('should render correctly', () => {
        const { asFragment } = renderWithProps();

        expect(asFragment()).toMatchSnapshot();
    });

    it('should render correctly with className', () => {
        renderWithProps({ className: 'foo' });

        expect(screen.getByTestId('navbar')).toHaveClass('foo');
    });
});

