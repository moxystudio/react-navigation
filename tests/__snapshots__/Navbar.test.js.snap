// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar should render correctly 1`] = `
<DocumentFragment>
  <div
    class="react-navigation__container top"
    data-testid="navbar"
  >
    <nav
      class="react-navigation__navbar top"
    >
      <div>
        <a
          href="/"
        >
          next-with-moxy
        </a>
      </div>
      <ul>
        <li>
          <a
            href="/lorem"
          >
            Lorem
          </a>
        </li>
        <li>
          <a
            href="/ipsum"
          >
            Ipsum
          </a>
        </li>
      </ul>
    </nav>
    <button>
      boom
    </button>
  </div>
</DocumentFragment>
`;
